refactor(metrics): remove Codahale add Micrometer

djordje-mijatovic opened the pull request:
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

szalai1 commented:
hey @djordje-mijatovic,

Thanks for contribution.

* Any change that changes exposed metrics at this level, needs to be feature flagged, so teams relaying on the metrics for alerting and dashboarding can rollout whenever they want.
* We agree that dropwizzard is not ideal, and planning to add proper prometheus support with labels, histrograms.

I'm keen know what are the problems you are solving with this change.

aditya-radhakrishnan commented:
Hey @djordje-mijatovic have you had a chance to take a look at Peter's feedback?

djordje-mijatovic commented:
> Hey @djordje-mijatovic have you had a chance to take a look at Peter's feedback?

Who is Peter? 

If you mean the comment above, what this PR solves... I have removed CodaHale and instead of it, I have added Micrometer. 
@mattmatravers I suppose you can provide more details why we did it.

mattmatravers commented:
> Who is Peter?
I think that's @szalai1.

> If you mean the comment above, what this PR solves... I have removed CodaHale and instead of it, I have added Micrometer.
@mattmatravers I suppose you can provide more details why we did it.

My understanding here is that CodeHale requires an extra JMX component to expose its metrics, and we'd like to reduce the overall API/vulnerability surface of DataHub. Micrometer is also the standard for metrics in Spring Boot so we'd rather not be using two standards where possible.

mattmatravers commented:
> * Any change that changes exposed metrics at this level, needs to be feature flagged, so teams relaying on the metrics for alerting and dashboarding can rollout whenever they want.

A feature flag would be ideal for backwards compatibility. How much more work would this involve, @djordje-mijatovic?

djordje-mijatovic commented:
> A feature flag would be ideal for backwards compatibility. How much more work would this involve, @djordje-mijatovic?

We can't do that as the whole CodaHale code is already removed. Reverting CodaHale code is an option, but it will have a very negative effect on the performance so I wouldn't go this way.

mattmatravers commented:
@djordje-mijatovic I think you'll need to merge changes from master to fix the red builds.

Going over this again, I have some suspicion that we might not have an exactly like for like replacement, e.g. LongTaskTimer has a very special way of working if I recall correctly. We could add unit tests to prove we're recording it correctly or add feature flags, but this would add a lot of extra effort.

@szalai1 do you have any suggestions for a way forward here? Ideally this would be a drop-in replacement I don't think there are any existing unit tests.

szalai1 commented on docker/monitoring/docker-compose.monitoring.yml:
```diff
@@ -1,35 +1,6 @@
-      - ENABLE_PROMETHEUS=true
-      - ENABLE_OTEL=true
-      - OTEL_TRACES_EXPORTER=jaeger
-      - OTEL_EXPORTER_JAEGER_ENDPOINT=http://jaeger-all-in-one:14250
```
Open telemetry is a separate agent we add during runtime to support jaeger/tracing. The metrics change shouldn't impact this.

szalai1 commented on metadata-utils/src/main/java/com/linkedin/metadata/utils/metrics/MetricUtils.java:
```diff
@@ -1,51 +1,83 @@
+    return TIMER_MAP.get(name);
+  }
+
+  public static DistributionSummary histogram(Class<?> klass, String... subNames) {
```
when I run `GMS` with this change locally, I couldn't see percentiles of the distribution other then `max`.

szalai1 replied:
currently we have: min, 50th, 75th, 90th, 95th, 98, 99th, 999th max

szalai1 commented on metadata-utils/src/main/java/com/linkedin/metadata/utils/metrics/MetricUtils.java:
```diff
@@ -1,51 +1,83 @@
 
-  public static Counter counter(String metricName) {
-    return REGISTRY.counter(MetricRegistry.name(metricName));
+  public static String buildName(Class<?> klass, String... subNames) {
```
Metrics names changed:
before:
`metrics_com_datahub_graphql_GraphQLController_searchAcrossEntities_Count` 
after: `class_com_datahub_graphql_GraphQLController_searchAcrossEntities_count`

szalai1 commented:
@mattmatravers @djordje-mijatovic 
yes, I'm Peter. 

> My understanding here is that CodeHale requires an extra JMX component to expose its metrics, and we'd like to reduce the overall API/vulnerability surface of DataHub. Micrometer is also the standard for metrics in Spring Boot so we'd rather not be using two standards where possible.

This makes sense. JMX also provides extra JVM metrics, which are (can be) used for alerting, monitoring. Do you use these metrics? If not you can disable monitoring in helm and the jmx exporter won't be loaded in runtime (it's not even needed in the container), if yes with this change they will disappear. 

This change is ok, if 
* it produces identical metrics (same metric names and labels) as before
* or change is behind a feature flag

Metrics diffs I observed:
* missing jmx metrics (jvm_info)
* names are structured differently (commented above)
* labels are missing (e.g type="histogram" )

Notes: 
* It's on our backlog to fully rewrite how we measure things in datahub, to fully utilise Prometheus. (e.g the current metrics calculate latency percentiles in the app, which should be done in Prometheus. ), it's more work than just replacing a lib and will break/fix metrics. 
* Changes on metric endpoint (port, path), needs to be updated in `datahub-helm` too ( I can do that)
* KafkaHealthChecker.java L40 needs to be updated too, builds fails 
* These dashboards supposed to work after the change: https://github.com/datahub-project/datahub/tree/master/docker/monitoring/grafana/dashboards
* Docs: https://github.com/datahub-project/datahub/tree/master/docker/monitoring/grafana/dashboards

szalai1 commented:
We talked about this with the team. Sharing our plans, so to get aligned. These are on our backlog, but not the highest priority at the moment: 
* we would like to use metrics native to Prometheus in the long-run with labels and distributed histograms. Probably we are going to use: https://github.com/prometheus/client_java , with/or [docs](https://docs.spring.io/spring-metrics/docs/current/public/prometheus), this requires reimplemented all metrics to assign appropriate labels and having naming conventions. 
* we would like to roll the change with flag to turn the new monitoring on and off, and not breaking the current way for those who use it in production. 
* keeping JMX exporter optional, because it's not required for internal metics, but provides manny valuable metrics about the JVM.

djordje-mijatovic commented:
> 

@szalai1 

Hello. Thank you for your comments. In the meantime we talk that it would be much safer to do it step by step so I have create another PR: https://github.com/datahub-project/datahub/pull/6889
This one should be more aligned and it should be testable. The Codahale is not removed but it is encapsulate. Only in the MetricUtils class you can see Codahale imports. In the same file I have added the Micrometer register and wrote MetricUtilsTest to comfirm both metric works the same way. Unfortunately, metric are a bit different so I had to introduce some modifications - everythin is documents in JavaDoc comments in MetricUtilsTest file.
In theory this PR should not effect on CodaHale metrics, it should just add Micrometer metrics too. Once you are satisfied with the results we could remove Codahale or add a feature switch and leave it.
